### PR TITLE
Added support for PHP 8.1

### LINKED ISSUE #19: Switch from Travis-Ci to GitHub actions for continuous integration workflow
To prepare a pull request following these steps:

- Remove the `CHANGELOG.md` file (we will be doing changelogs in the milestone descriptions, and those get propagated to tags and release notes)
- Add the workflow file:
  ```bash
  mkdir -p .github/workflows ; cd  .github/workflows ; wget https://gist.githubusercontent.com/weierophinney/9decd19f76b7d9745c6559074053fa65/raw/6ffb33e59796cfec569405139aa65da9396ea5cd/continuous-integration.yml
  ```
- Check to see if there are any differences between what the Travis config and the GHA workflow do - don't worry about code coverage - but check for things like: extensions required, any pre-script tasks, etc. (Most of the time, there aren't any). If you see any, either: (1) add a `.laminas-ci.json` file with required configuration (e.g., adding extensions, adding php.ini settings; see https://github.com/laminas/laminas-ci-matrix-action#configuration for details), or (2) add information to the GHA workflow file (e.g., adding a service such as MySQL, Redis, etc.; see the laminas-db or laminas-cache adapter packages for examples).
- Remove the `.travis.yml` file.
- Remove the entry for `composer.lock` from the `.gitignore` file.
- In the `.gitattributes` file:
  - Remove the entry for `.travis.yml`
  - Add an entry for `.laminas-ci.json` if you created one
  - Add an entry for `.github` if one does not exist
- Using PHP 7.4, run `composer update`
- Update the `README.md` file:
  - Remove any existing badges for Travis and/or Coveralls.
  - Add a badge with the following (substituting in the correct organization and repository name):
    ```markdown
    [![Build Status](https://github.com/laminas/laminas-ldap/actions/workflows/continuous-integration.yml/badge.svg)](https://github.com/laminas/laminas-ldap/actions/workflows/continuous-integration.yml)
    ```
- Add and commit the GHA workflow, the `.gitignore` and `.gitattribute` files, the `composer.lock` file, and the `README.md` file.
- Push to your fork and create the PR. If there are any QA job failures, fix and push the changes.
- If the package supports PHP `^5.6 | ^7.0`, you will need to exclude locked runs for PHP 5.6 - 7.2. The way to do that is to create the `.laminas-ci.json` file if you haven't already, and ensure it has the following contents:
  ```json
  {
      "exclude": [
          {"name": "PHPUnit on PHP 5.6 with locked dependencies"},
          {"name": "PHPUnit on PHP 7.0 with locked dependencies"},
          {"name": "PHPUnit on PHP 7.1 with locked dependencies"},
          {"name": "PHPUnit on PHP 7.2 with locked dependencies"}
      ]
  }
  ```

### LINKED ISSUE #22: PHP 8.1 support
### Feature Request

|    Q        |   A
|------------ | ------
| New Feature | yes

#### Summary

PHP with [version 8.1 is released](https://www.php.net/releases/8.1/index.php) and for this component the support of the new version is still missing.

### PR DESCRIPTION
**Pull request by Koen1999**
|    Q          |   A
|-------------- | ------
| Documentation | no
| Bugfix        | no
| BC Break      | no
| New Feature   | yes
| RFC           | no
| QA            | no

### Description
This PR adds support for PHP 8.1. No big changes had to be made luckily. I ran a code inspection indicating no compatibility issues. The tests that I could easily run all passed. Note that Laminas_Ldap online tests were not available.
This fixes #22

### CODE REVIEW

**Comment by Koen1999** [comment]
Travis builds do not support PHP 8.1 yet, so I did not include them.

**Comment by snapshotpl** [comment]
#19 should be done before

**Comment by snapshotpl** [comment]
Looks like #20 try to resolve #19

**Comment by Koen1999** [comment]
> Looks like #20 try to resolve #19

Sadly, I know too little about Github actions to assist with this.

**Comment by froschdesign** [comment]
@Koen1999 
> Travis builds do not support PHP 8.1 yet, so I did not include them.

Is there a nightly build or something else we can use here?

**Comment by Koen1999** [comment]
> @Koen1999
> 
> > Travis builds do not support PHP 8.1 yet, so I did not include them.
> 
> Is there a nightly build or something else we can use here?

I tried it, but nightly is still on 8.0.3

**Comment by Ocramius** [comment]
> I know too little about Github actions to assist with this.

Give it a shot: merge #20 into this branch, and it should run CI.

We can't proceed without CI, and Travis-CI is dead-dead-dead.

**Review by Ocramius — CHANGES_REQUESTED** [review]
If we fix the few issues in `docs/book/api.md`, we should be good here :thinking:

File: composer.json
Comment (Ocramius) [inline composer.json]: This dependency should be dropped too: it's a relatively simple change, please see https://github.com/laminas/technical-steering-committee/blob/c7f6babfb1773d9cb9371208102cbb8b27980bd5/meetings/minutes/2021-08-02-TSC-Minutes.md#remove-laminaslaminas-zendframework-bridge-dependency-from-our-packages

**Review by Ocramius — CHANGES_REQUESTED** [review]

**Comment by froschdesign** [comment]
@Koen1999 
The updates for the documentation:

From:

```markdown
- `$scope`: The search scope:
    - `Ldap::SEARCH_SCOPE_SUB` searches the complete subtree including the
      `$baseDn` node. This is the default value.
    - `Ldap::SEARCH_SCOPE_ONE` restricts search to one level below `$baseDn`.
    - `Ldap::SEARCH_SCOPE_BASE` restricts search to the `$baseDn` itself; this
      can be used to efficiently retrieve a single entry by its DN.
```

to:

```markdown
- `$scope`: The search scope:
  - `Ldap::SEARCH_SCOPE_SUB` searches the complete subtree including the
    `$baseDn` node. This is the default value.
  - `Ldap::SEARCH_SCOPE_ONE` restricts search to one level below `$baseDn`.
  - `Ldap::SEARCH_SCOPE_BASE` restricts search to the `$baseDn` itself; this
    can be used to efficiently retrieve a single entry by its DN.
```

https://github.com/laminas/laminas-ldap/blob/9996468cb0a5b313b49c5e33885d526c5f89ebe9/docs/book/api.md?plain=1#L94-L97

From:
```markdown
static setDateTimeAttribute(/* ... */) : void                                              | See [setDateTimeAttribute()](#setdatetimeattribute) below for details.
static getDateTimeAttribute(/* ... */) : array|int                                         | See [getDateTimeAttribute()](#getdatetimeattribute) below for details.

```
to:
```markdown
`static setDateTimeAttribute(/* ... */) : void`                                            | See [setDateTimeAttribute()](#setdatetimeattribute) below for details.
`static getDateTimeAttribute(/* ... */) : array|int`                                       | See [getDateTimeAttribute()](#getdatetimeattribute) below for details.

```

https://github.com/laminas/laminas-ldap/blob/9996468cb0a5b313b49c5e33885d526c5f89ebe9/docs/book/api.md?plain=1#L170-L171

File: composer.json
Comment (Koen1999) [reply composer.json]: It is still in the lockfile because of:
```
php composer.phar why laminas/laminas-zendframework-bridge
laminas/laminas-config        3.5.0  requires  laminas/laminas-zendframework-bridge (^1.0)  
laminas/laminas-eventmanager  3.3.1  requires  laminas/laminas-zendframework-bridge (^1.0)  
laminas/laminas-stdlib        3.3.1  requires  laminas/laminas-zendframework-bridge (^1.0)  
```

Should I bump the versions of those packages to completely remove the dependency?

**Comment by froschdesign** [comment]
@Koen1999 
Remove all the contents from the file `docs/book/index.md`, but the file itself must still be there.

File: composer.json
Comment (Ocramius) [reply composer.json]: No, it's fine - at least we got rid of the dependency from here :-) 

Further upgrades will slowly phase it out everywhere :+1:

**Review by Ocramius — CHANGES_REQUESTED** [review]
```
PHP Fatal error:  During inheritance of Countable: Uncaught Return type of Laminas\Config\Config::count() should either be compatible with Countable::count(): int, or the #[\ReturnTypeWillChange] attribute should be used to temporarily suppress the notice
```

Probably needs an upgrade of `laminas/config`, or complete phase out of the dependency, which seems to only be in use in tests

**Comment by Koen1999** [comment]
> ```
> PHP Fatal error:  During inheritance of Countable: Uncaught Return type of Laminas\Config\Config::count() should either be compatible with Countable::count(): int, or the #[\ReturnTypeWillChange] attribute should be used to temporarily suppress the notice
> ```
> 
> Probably needs an upgrade of `laminas/config`, or complete phase out of the dependency, which seems to only be in use in tests

I'll bump it to ^3.7.0

**Comment by Koen1999** [comment]
To me it seems as if the remaining issues in the failings build are due to missing support in PHPUnit and laminas-config.

**Comment by Ocramius** [comment]
PHPUnit could need a dependency upgrade: the warnings are valid (need fixing), and caused by `null` passed to parameters that don't accept it: https://wiki.php.net/rfc/deprecate_null_to_scalar_internal_arg

**Comment by froschdesign** [comment]
@Koen1999 
An update for the `phpcs.xml` file is needed:

```xml
<?xml version="1.0"?>
<ruleset xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
         xsi:noNamespaceSchemaLocation="./vendor/squizlabs/php_codesniffer/phpcs.xsd">

    <arg name="basepath" value="."/>
    <arg name="cache" value=".phpcs-cache"/>
    <arg name="colors"/>
    <arg name="extensions" value="php"/>
    <arg name="parallel" value="80"/>

    <!-- Show progress -->
    <arg value="p"/>

    <!-- Paths to check -->
    <file>src</file>
    <file>test</file>

    <!-- Include all rules from Laminas Coding Standard -->
    <rule ref="LaminasCodingStandard"/>
</ruleset>
```

**Comment by Ocramius** [comment]
Oooof, PHPCS failing on this is annoying, and may be a long rabbit hole :thinking: 

Not sure how to proceed - ideally would want to go from green to green here /cc @heiglandreas

**Comment by Koen1999** [comment]
The problem is that PHPCS officially only supports PhpUnit 7. But for PhpUnit 7 (which isn't supported anymore anyway), our tests fail because of incompatibility with PHP 8.1. I propose allowing this release for now and hoping PHPCBF makes progress towards supporting newer PhpUnit versions soon. If I had to choose between failing tests and failing style checks, I would happily accept the failing style checks.

**Comment by froschdesign** [comment]
@Koen1999 
I do not know what you mean but check with PHP_CodeSniffer works. See:

https://github.com/laminas/laminas-ldap/runs/4381682162?check_suite_focus=true#step:3:458

The problem is that there are more than 3000 errors and warnings!

---

The main idea was to add PHP 8.1 support and not to fix the coding style. We have package with support for PHP with version 8.1 and laminas-coding-standard with version 1. Look at [laminas-view](https://github.com/laminas/laminas-view/blob/2a5ce1b6dbaf2f965bad65852e309d6b7e8db4e2/composer.json#L21-L30).

**Comment by froschdesign** [comment]
> @Koen1999 I do not know what you mean but check with PHP_CodeSniffer works. See:
> 
> https://github.com/laminas/laminas-ldap/runs/4381682162?check_suite_focus=true#step:3:458
> 
> The problem is that there are more than 3000 errors and warnings!
> 
> The main idea was to add PHP 8.1 support and not to fix the coding style. We have packages with support for PHP with version 8.1 and laminas-coding-standard with version 1. Look at [laminas-view](https://github.com/laminas/laminas-view/blob/2a5ce1b6dbaf2f965bad65852e309d6b7e8db4e2/composer.json#L21-L30).

**Comment by Koen1999** [comment]
I can revert the upgrade to coding standard 2.2 which was done in the other PR if that solves the issue?

**Comment by froschdesign** [comment]
> I can revert the upgrade to coding standard 2.2 which was done in the other PR if that solves the issue?

The other option would be to fix all 3000 errors and warnings, but I don't want to do that to you.
So far it has worked for other components, so I hope it works here too.

**Comment by Koen1999** [comment]
This is an odd one that we didn't have before :/
https://github.com/laminas/laminas-ldap/runs/4387798576?check_suite_focus=true#step:3:114

**Comment by froschdesign** [comment]
@Koen1999 
I suggest to check laminas-view or laminas-i18n or laminas-config as reference. The following was added there:

```json
{
    "ignore_php_platform_requirements": {
        "8.1": true
    }
}
```

https://github.com/laminas/laminas-config/blob/3.8.x/.laminas-ci.json

**Comment by froschdesign** [comment]
> This is an odd one that we didn't have before :/ https://github.com/laminas/laminas-ldap/runs/4387798576?check_suite_focus=true#step:3:114

😞 

The following message is thrown:

> Cannot install extensions for the current PHP version.
Please use ".laminas-ci/pre-run.sh" to setup specific extensions for PHP 8.1
Additional details can be found on https://stackoverflow.com/q/8141407
The following extensions were not installed: ldap

**Comment by Koen1999** [comment]
I think the upgrade of Laminas coding standard was neccessary in order to run PHPCS on PHP 7.4 without deprecation warnings.

**Comment by Koen1999** [comment]
PHPCS isn't failing anymore 🥳 

Regarding the failing QA checks for PHP 8.1, can we agree this is a bug and will likely be fixed later? Moreover, the build for PHP 8.1 was passing earlier and no functional changes were made since.

**Comment by froschdesign** [comment]
> …can we agree this is a bug and will likely be fixed later…

All checks must passed also the [DCO](https://github.com/laminas/laminas-ldap/pull/23/checks?check_run_id=4388166572)!

And to solve the _hopefully_ last problem, the note should be followed:

> Cannot install extensions for the current PHP version.
Please use ".laminas-ci/pre-run.sh" to setup specific extensions for PHP 8.1
Additional details can be found on https://stackoverflow.com/q/8141407
The following extensions were not installed: ldap 

https://github.com/laminas/laminas-ldap/runs/4388195315?check_suite_focus=true#step:3:18

**Comment by Koen1999** [comment]
> > …can we agree this is a bug and will likely be fixed later…
> 
> All checks must passed also the [DCO](https://github.com/laminas/laminas-ldap/pull/23/checks?check_run_id=4388166572)!
> 
> And to solve the _hopefully_ last problem, the note should be followed:
> 
> > Cannot install extensions for the current PHP version.
> > Please use ".laminas-ci/pre-run.sh" to setup specific extensions for PHP 8.1
> > Additional details can be found on https://stackoverflow.com/q/8141407
> > The following extensions were not installed: ldap
> 
> https://github.com/laminas/laminas-ldap/runs/4388195315?check_suite_focus=true#step:3:18

The DCO did not pass in #20 which was suggested to merge into this PR. I do not see how I can make their commits adhere to the DCO.

I'll look into the second issue.

**Comment by Koen1999** [comment]
The commit missing the Signoff is the Laminas-Bot :/

We can merge into 2.11.x

Or....

**Comment by Ocramius** [comment]
It's fine: I'll tell DCO to shut up 👍

**Comment by Koen1999** [comment]
I'll see if I can get https://github.com/laminas/laminas-authentication/pull/24 to work after this is merged and released.

**Review by snapshotpl — APPROVED** [review]

**Comment by snapshotpl** [comment]
@Ocramius anything to add?

**Review by Ocramius — APPROVED** [review]
:ship: thanks @Koen1999

**Comment by kaufmajo** [comment]
It seems that this package is missing that since  php 8.1, the return type for "ldap_connect" returns an LDAP\Connection instance now - previously, a resource was returned. 

Screenshot 2022-01-13 084800

**Comment by Ocramius** [comment]
@kaufmajo I'd suggest:

 1. adding tests for scenarios that currently crash because of PHP 8.1 breaking changes (none found by existing test suite :( )
 2. help in reducing the psalm baseline @ https://github.com/laminas/laminas-ldap/blob/5e0debb0c2e38f5c197eacf7a8d99b79f74e1110/psalm-baseline.xml - specifically focus on PHP 8.1 analysis. 

Patches welcome, so get cracking :)

**Comment by kaufmajo** [comment]
@Ocramius  Be assured, I would love to, but i am having a lack of knowledge how to write tests and provide patches with github. It's still on my list.

**Comment by Ocramius** [comment]
Give it a shot when you can - package is not going anywhere :D

**Pull request merged**
